pkg/readiness/object_tracker_test.go

```diff
@@ -243,3 +244,64 @@ func Test_ObjectTracker_kinds(t *testing.T) {
 	g.Expect(kindsBefore).ShouldNot(gomega.BeEmpty(), "expected non-empty kinds")
 	g.Expect(kindsAfter).Should(gomega.Equal(kindsBefore), "expected kinds to match")
 }
+
+// Verify that TryCancelExpect functions the same as regular CancelExpect if readinessRetries is set to 0
+func Test_ObjectTracker_TryCancelExpect_Default(t *testing.T) {
+	g := gomega.NewWithT(t)
+	ot := newObjTracker(schema.GroupVersionKind{}, func() objData {
+		return objData{retries: 0}
+	})
+
+	const count = 10
+	ct := makeCTSlice("ct-", count)
+	for i := 0; i < len(ct); i++ {
+		ot.Expect(ct[i])
+	}
+	g.Expect(ot.Satisfied()).NotTo(gomega.BeTrue(), "should not be satisfied before ExpectationsDone")
+	ot.ExpectationsDone()
+
+	// Skip the first two
+	for i := 2; i < len(ct); i++ {
+		g.Expect(ot.Satisfied()).NotTo(gomega.BeTrue(), "should not be satisfied before observations are done")
+		ot.Observe(ct[i])
+	}
+	g.Expect(ot.Satisfied()).NotTo(gomega.BeTrue(), "two expectation remain")
+
+	ot.TryCancelExpect(ct[0])
+	g.Expect(ot.Satisfied()).NotTo(gomega.BeTrue(), "one expectation remains")
+
+	ot.TryCancelExpect(ct[1])
+	g.Expect(ot.Satisfied()).To(gomega.BeTrue(), "should be satisfied")
+}
+
+// Verify that TryCancelExpect must be called multiple times before an expectation is cancelled
+func Test_ObjectTracker_TryCancelExpect_WithRetries(t *testing.T) {
+	g := gomega.NewWithT(t)
+	ot := newObjTracker(schema.GroupVersionKind{}, func() objData {
+		return objData{retries: 2}
+	})
+
+	const count = 10
+	ct := makeCTSlice("ct-", count)
+	for i := 0; i < len(ct); i++ {
+		ot.Expect(ct[i])
+	}
+	g.Expect(ot.Satisfied()).NotTo(gomega.BeTrue(), "should not be satisfied before ExpectationsDone")
+	ot.ExpectationsDone()
+
+	// Skip the first one
+	for i := 1; i < len(ct); i++ {
+		g.Expect(ot.Satisfied()).NotTo(gomega.BeTrue(), "should not be satisfied before observations are done")
+		ot.Observe(ct[i])
+	}
+	g.Expect(ot.Satisfied()).NotTo(gomega.BeTrue(), "one expectation remains with two retries")
+
+	ot.TryCancelExpect(ct[0])
+	g.Expect(ot.Satisfied()).NotTo(gomega.BeTrue(), "one expectation remains with one retries")
+
+	ot.TryCancelExpect(ct[0])
+	g.Expect(ot.Satisfied()).NotTo(gomega.BeTrue(), "one expectation remains with zero retries")
+
+	ot.TryCancelExpect(ct[0])
+	g.Expect(ot.Satisfied()).To(gomega.BeTrue(), "should be satisfied")
+}
```

Comment:
We should test the infinite behavior too...

I realize that we can't test for infinite retries, but if we set retry count to -1, cancel, and retry count is still -1 that implies that we will retry indefinitely.

Reply:
I added a test that sets `retries` to `-1` and then does 20 `TryCancelExpect` calls.  This was the best I could think of as far as testing the `.*Expect()` functions.

The test you're suggesting seems to be more of a test of the `objData` struct itself.  Happy to add that, just didn't because it seemed to break from the existing testing pattern.

Thoughts?

Reply:
IMO the 20x test is sufficient